py_zipkin/request_helpers.py

```diff
@@ -0,0 +1,147 @@
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
+                                decision, false-y value otherwise
+    """
+    parsed = {
+        'trace_id': headers.get('X-B3-TraceId', None),
+        'span_id': headers.get('X-B3-SpanId', None),
+        'parent_span_id': headers.get('X-B3-ParentSpanId', None),
+        'sampled_str': headers.get('X-B3-Sampled', None),
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
+    # non-empty, but invalid string.
+    if parsed['sampled_str'] == '':
+        parsed['sampled_str'] = 'INVALID'
+    parsed['sample_decision_only'] = (
+        all(parsed[k] is None for k in ('trace_id', 'span_id', 'parent_span_id'))
+        and parsed['sampled_str'])
+    return parsed
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
+                                      use_128bit_trace_id=False):
+    """
+    Implements extraction of B3 headers per:
+        https://github.com/openzipkin/b3-propagation
+
+    The input headers can be any dict-like container that supports "in"
+    membership test and a .get() method that accepts a default value.
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        parsed = _parse_single_header(headers['b3'])
+    else:
+        parsed = _parse_multi_header(headers)
+
+    # Handle the lone-sampling-decision case:
+    if parsed['sample_decision_only']:
+        # See if it's a valid encoding of (only) sample decision
+        if parsed['sampled_str'] in ('0', 'false'):
+            sample_rate = 0.0
+        elif parsed['sampled_str'] in ('1', 'true', 'd'):
+            sample_rate = 100.0
+        else:
+            # Nonsense gets a None
+            log.warning('Ignoring invalid sampling decision %r; '
+                        'requires "0", "1", or "d".', parsed['sampled_str'])
+            return None
+        attrs = create_attrs_for_span(
+            sample_rate=sample_rate,
+            use_128bit_trace_id=use_128bit_trace_id,
+            flags='1' if parsed['sampled_str'] == 'd' else '0',
+        )
+        return attrs
+
+    # Enforce requirement of having trace & span IDs
+    if not parsed['trace_id'] or not parsed['span_id']:
+        log.warning('Ignoring invalid or missing TraceId & SpanId:'
+                    '%r, %r', parsed['trace_id'], parsed['span_id'])
+        return None
+
+    # Handle sampling decision
+    if parsed['sampled_str']:
```

Comment:
not sure why we are doing this again.. similar comment to elsewhere.. be tight where the actual parsing occurs (ex you can log the input there if needed). then you don't need to do it again later.

Reply:
I'm sensing a theme about handing bad input in the parsing, perhaps with exceptions for flow control, simplifying the common code post-parsing...